lib/pure/ioselectors.nim

```diff
@@ -0,0 +1,1768 @@
+      result.maxFD = maxFD.uint
+      result.fds = allocSharedArray[SelectorKey[T]](maxFD)
+      result.pollfds = allocSharedArray[TPollFd](maxFD)
+      when hasThreadSupport:
+        initLock(result.lock)
+
+    proc close*[T](s: Selector[T]) =
+      when hasThreadSupport:
+        deinitLock(s.lock)
+      deallocSharedArray(s.fds)
+      deallocSharedArray(s.pollfds)
+      deallocShared(cast[pointer](s))
+
+    template pollAdd[T](s: Selector[T], sock: cint, events: set[Event]) =
+      withPollLock(s):
+        var pollev: cshort = 0
+        if Event.Read in events: pollev = pollev or POLLIN
+        if Event.Write in events: pollev = pollev or POLLOUT
+        s.pollfds[s.pollcnt].fd = cint(sock)
+        s.pollfds[s.pollcnt].events = pollev
+        inc(s.count)
+        inc(s.pollcnt)
+
+    template pollUpdate[T](s: Selector[T], sock: cint, events: set[Event]) =
+      withPollLock(s):
+        var i = 0
+        var pollev: cshort = 0
+        if Event.Read in events: pollev = pollev or POLLIN
+        if Event.Write in events: pollev = pollev or POLLOUT
+
+        while i < s.pollcnt:
+          if s.pollfds[i].fd == sock:
+            s.pollfds[i].events = pollev
+            break
+          inc(i)
+
+        if i == s.pollcnt:
+          raise newException(ValueError,
+                             "Descriptor is not registered in queue")
+
+    template pollRemove[T](s: Selector[T], sock: cint) =
+      withPollLock(s):
+        var i = 0
+        while i < s.pollcnt:
+          if s.pollfds[i].fd == sock:
+            if i == s.pollcnt - 1:
+              s.pollfds[i].fd = 0
+              s.pollfds[i].events = 0
+              s.pollfds[i].revents = 0
+            else:
+              while i < (s.pollcnt - 1):
+                s.pollfds[i].fd = s.pollfds[i + 1].fd
+                s.pollfds[i].events = s.pollfds[i + 1].events
+                inc(i)
+            break
+          inc(i)
+        dec(s.pollcnt)
+        dec(s.count)
+
+    proc registerHandle*[T](s: Selector[T], fd: SocketHandle,
+                            events: set[Event], data: T) =
+      var fdi = int(fd)
+      s.checkMaxFd(fdi)
+      doAssert(s.fds[fdi].ident == 0)
+      setKey(s, fdi, fdi, {Event.flagHandle} + events, 0, data)
+      s.pollAdd(fdi.cint, events)
+
+    proc updateHandle*[T](s: Selector[T], fd: SocketHandle,
+                          events: set[Event]) =
+      var fdi = int(fd)
+      s.checkMaxFd(fdi)
+      var oe = s.fds[fdi].flags
+      doAssert(s.fds[fdi].ident != 0)
+      doAssert(Event.flagHandle in oe)
+      var ne = events + {Event.flagHandle}
+      if ne != oe:
+        if events != {}:
+          s.pollUpdate(fd.cint, events)
+        else:
+          s.pollRemove(fd.cint)
+        s.fds[fdi].flags = ne
+
+    proc registerTimer*[T](s: Selector[T], timeout: int, oneshot: bool,
+                           data: T): int {.discardable.} =
+      raise newException(ValueError, "Not implemented")
```

Comment:
what? don't have procs that are "not implemented". we prefer compile-time errors.